ONSdigital/sdc-cryptography
# Use pipenv, improve unit test coverage and unpin dependencies

## What? and Why?
https://trello.com/c/YdQbbZdp/2551-spring-clean-on-sdc-cryptography

This PR increases the unit test coverage, unpins the dependency versions and uses Pipenv for package management whilst developing.

In setup.py, it used to get its install_requires field from the requirements file.  Now that we have a pipfile instead, I wasn't sure of the best way to do this.  Hard coding the 3 dependencies works, but I doubt it's the best way of doing it but I couldn't find a decent solution.

The unit test coverage was vastly improved, but testing the encrypt function was difficult as the same input gave a different output each time.  I couldn't figure out how to get around that.

## How to test
- Check out this repo and `pipenv run make sdist`
- `cd sdx-decrypt` (wherever it is on your machine)
- After sorting out the virtual environment, `make build` then `pip install <path to sdc_cryptography>/sdc-cryptography-0.3.0.tar.gz` to install the new version you just created.
- Spin up a standalone sdx-decrypt`export PORT=5000; python server.py` 
- Spin up sdx-compose.
- Stop the sdx-gateway service. `docker ps` then `docker stop <gateway container id>`
- Submit a submission via sdx-console.  It should now be on the rabbit queue, waiting to be consumed.
- Go to localhost:15672 for the management console (rabbit / rabbit).  Get the message you just submitted, then POST it to the decrypt service (localhost:5000/decrypt) you have running locally (I used Postman).  It should decrypt and give you the expected payload that you submitted via console.

## Checklist
